File: src/ios/CDVCardFlight.m
#import "CDVCardFlight.h"

@implementation CDVCardFlight

@synthesize readerAttachedCallbackID;
@synthesize readerConnectingCallbackID;
@synthesize readerSwipeCallbackID;
@synthesize cardReadCallbackID;
@synthesize readerBatteryLowCallbackID;
@synthesize readerUndetectedCallbackID;
@synthesize readerConnectedCallbackID;
@synthesize readerDetatchedCallbackID;
@synthesize currentCallbackID;


- (CDVPlugin *)initWithWebView:(UIWebView *)theWebView {
  self = (CDVCardFlight *)[super initWithWebView:theWebView];
  _cardFlightAuthorized = NO;
  return self;
}

- (void)readerIsAttached {
  if (self.readerAttachedCallbackID) {
    [self sendSuccessMessage:@"Reader is attached" withCallbackID:self.readerAttachedCallbackID];
  }
}

- (void)readerIsConnecting {
  if (self.readerConnectingCallbackID) {
    [self sendSuccessMessage:@"Reader is connecting" withCallbackID:self.readerConnectingCallbackID];
  }
}

- (void)readerSwipeDetected {
  if (self.readerSwipeCallbackID) {
    [self sendSuccessMessage:@"Swipe detected" withCallbackID:self.readerSwipeCallbackID];
  }
}

- (void)readerBatteryLow {
  if (self.readerBatteryLowCallbackID) {
    [self sendSuccessMessage:@"Reader battery is low" withCallbackID:self.readerBatteryLowCallbackID];
  }
}

- (void)readerNotDetected {
  if (self.readerUndetectedCallbackID) {
    [self sendSuccessMessage:@"Reader not detected" withCallbackID:self.readerUndetectedCallbackID];
  }
}

- (void)readerIsConnected:(BOOL)isConnected withError:(NSError *)error {
  if (!self.readerConnectedCallbackID) {
    return NSLog(@"No readerConnectedCallbackID to trigger");
  }
  
  if (error) {
    [self sendErrorMessage:[error localizedDescription] withCallbackID:self.readerConnectedCallbackID];
  } else if (isConnected) {
    [self sendSuccessMessage:@"Reader connected" withCallbackID:self.readerConnectedCallbackID];
  } else {
    [self sendErrorMessage:@"Reader is not connected, no error given" withCallbackID:self.readerConnectedCallbackID];
  }
}

- (void)transactionResult:(CFTCharge *)charge withError:(NSError *)error {
  if (error) {
    NSLog(@"transactionResult callback with error: %@", error);
  } else {
    // TODO: handle success
  }
}


- (void)readerIsDisconnected {
  if (self.readerDetatchedCallbackID) {
    [self sendSuccessMessage:@"Reader disconnected" withCallbackID:self.readerDetatchedCallbackID];
  }
}

- (void)onReaderAttached:(CDVInvokedUrlCommand*)command {
  [self setReaderAttachedCallbackID:[command callbackId]];
}

- (void)onReaderConnecting:(CDVInvokedUrlCommand*)command {
  [self setReaderConnectingCallbackID:[command callbackId]];
}

- (void)onSwipeDetected:(CDVInvokedUrlCommand*)command {
  [self setReaderSwipeCallbackID:[command callbackId]];
}

- (void)onCardRead:(CDVInvokedUrlCommand*)command {
  [self setCardReadCallbackID:[command callbackId]];
}

- (void)onBatteryLow:(CDVInvokedUrlCommand*)command {
  [self setReaderBatteryLowCallbackID:[command callbackId]];
}

- (void)onReaderNotDetected:(CDVInvokedUrlCommand*)command {
  [self setReaderUndetectedCallbackID:[command callbackId]];
}

- (void)onReaderConnected:(CDVInvokedUrlCommand*)command {
  [self setReaderConnectedCallbackID:[command callbackId]];
}

- (void)onReaderDisconnected:(CDVInvokedUrlCommand*)command {
  [self setReaderDetatchedCallbackID:[command callbackId]];
}

- (void)authorizeCardFlightAccount:(CDVInvokedUrlCommand*)command {
  [self setCurrentCallbackID:[command callbackId]];

  NSString* cardFlightAPIToken = [command.arguments objectAtIndex:0];
  NSString* stripeMerchantToken = [command.arguments objectAtIndex:1];

  if (!cardFlightAPIToken) {
    return [self sendErrorMessage:@"Need to pass a CardFlight API token to initialize CardFlight"];
  }

  if (!stripeMerchantToken) {
    return [self sendErrorMessage:@"Need to pass a Stripe Merchant token to initialize CardFlight"];
  }

  [[CFTSessionManager sharedInstance] setApiToken:cardFlightAPIToken
                                      accountToken:stripeMerchantToken];

  _cardFlightAuthorized = YES;

  [self sendSuccessMessage:@"CardFlight authorized on this device"];
}

- (void)initializeReader:(CDVInvokedUrlCommand*)command {
  [self setCurrentCallbackID:[command callbackId]];

  _reader = [[CFTReader alloc] initWithReader:0];
  [_reader setDelegate:self];

  [self sendSuccessMessage:@"CardFlight reader ready"];
}

- (void)watchForSwipe:(CDVInvokedUrlCommand*)command {
  [self setCurrentCallbackID:[command callbackId]];

  [_reader beginSwipe];
  [self sendSuccessMessage:@"CardFlight reader awaiting swipe"];
}

- (void)readerCardResponse:(CFTCard *)card withError:(NSError *)error {
  if (error) {
    [self sendErrorMessage:[error localizedDescription] withCallbackID:self.cardReadCallbackID];
  } else {
    _card = card;
    [self sendSuccessMessage:@"Card read successfully" withCallbackID:self.cardReadCallbackID];

    // NSDictionary* dict = @{
    //   @"message": @"Card read successfully",
    //   @"cardName": [[NSString alloc] initWithString:card.name],
    //   @"cardNumber": [[NSString alloc] initWithString:card.encryptedSwipedCardNumber],
    //   @"cardExpirationMonth": [[NSString alloc] initWithString:card.expirationMonth],
    //   @"cardExpirationYear": [[NSString alloc] initWithString:card.expirationYear]
    // };

    // NSError *error; 
    // NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict 
    //                                         options:0
    //                                         error:&error];

    // if (!jsonData) {
    //   if (error) {
    //     [self sendErrorMessage:[error localizedDescription] withCallbackID:cardReadCallbackID];
    //   } else {
    //     [self sendErrorMessage:@"No json data found" withCallbackID:cardReadCallbackID];
    //   }
    // } else {
    //   NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    //   NSLog(@"Returning json: %@", jsonString);
    //   [self sendSuccessMessage:jsonString withCallbackID:self.cardReadCallbackID];
    // }
  }
}

- (void)tokenizeLastSwipe:(CDVInvokedUrlCommand*)command {
  NSString* tokenizeCallbackID = [command callbackId];

  if (!_card) {
    return [self sendErrorMessage:@"No card to tokenize" withCallbackID:tokenizeCallbackID];
  }

  [_card tokenizeCardWithSuccess:^{
      // NSDictionary* dict = @{
      //   @"message": @"Card tokenized with Stripe successfully",
      //   @"cardName": [[NSString alloc] initWithString:_card.name],
      //   @"cardNumber": [[NSString alloc] initWithString:_card.encryptedSwipedCardNumber],
      //   @"cardExpirationMonth": [[NSString alloc] initWithString:_card.expirationMonth],
      //   @"cardExpirationYear": [[NSString alloc] initWithString:_card.expirationYear],
      //   @"token": [[NSString alloc] initWithString: _card.cardToken]
      //   // @"cardName": _card.name,
      //   // @"cardNumber": _card.encryptedSwipedCardNumber,
      //   // @"cardExpirationMonth": _card.expirationMonth,
      //   // @"cardExpirationYear": _card.expirationYear,
      //   // @"token": _card.cardToken
      // };
      // [self sendSuccessDictionary:dict withCallbackID:tokenizeCallbackID];
      [self sendSuccessMessage:_card.cardToken withCallbackID:tokenizeCallbackID];
    }
    failure:^(NSError *error){
      [self sendErrorMessage:[error localizedDescription] withCallbackID:tokenizeCallbackID];
    }];
}

- (void)sendSuccessDictionary:(NSDictionary *)responseBody {
  CDVPluginResult* goingToCauseError = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:responseBody];
  CDVPluginResult* safe = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[goingToCauseError argumentsAsJSON]];
  [self.commandDelegate sendPluginResult:safe callbackId:self.currentCallbackID];
}

- (void)sendSuccessDictionary:(NSDictionary *)responseBody withCallbackID:(NSString *)callbackID {
  CDVPluginResult* goingToCauseError = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:responseBody];
  CDVPluginResult* safe = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[goingToCauseError argumentsAsJSON]];
  [self.commandDelegate sendPluginResult:safe callbackId:callbackID];
}

- (void)sendSuccessMessage:(NSString *)responseString {
  CDVPluginResult* successResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:responseString];
  [self.commandDelegate sendPluginResult:successResult callbackId:self.currentCallbackID];
}

- (void)sendSuccessMessage:(NSString *)responseString withCallbackID:(NSString *)callbackID {
  CDVPluginResult* successResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:responseString];
  [self.commandDelegate sendPluginResult:successResult callbackId:callbackID];
}

- (void)sendErrorMessage:(NSString *)errorMessage {
  CDVPluginResult* errorResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errorMessage];
  [self.commandDelegate sendPluginResult:errorResult callbackId:self.currentCallbackID];
}

- (void)sendErrorMessage:(NSString *)errorMessage withCallbackID:(NSString *)callbackID {
  CDVPluginResult* errorResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errorMessage];
  [self.commandDelegate sendPluginResult:errorResult callbackId:callbackID];
}

@end

File: src/android/CDVCardFlight.java
package org.weeels.plugins.cardflight;

import android.util.Log;
import com.getcardflight.models.CardFlight;
import com.getcardflight.models.Reader;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;

public class CDVCardFlight extends CordovaPlugin {

  private Reader reader;
  private CardFlightHandler handler;
  private CordovaInterface cdv;

  @Override
  public void initialize(CordovaInterface cordova, CordovaWebView webView) {
    super.initialize(cordova, webView);
    handler = new CardFlightHandler(cordova);
    cdv = cordova;
  }

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    boolean success = true;

    if (action.equals("authorizeCardFlightAccount")) {
      this.authorizeCardFlightAccount(args.getString(0), args.getString(1), callbackContext);
    } else if (action.equals("initializeReader")) {
      this.initializeReader(callbackContext);
    } else if (action.equals("onReaderAttached")) {
      handler.onReaderConnected(callbackContext);
    } else if (action.equals("onReaderConnecting")) {
      handler.onReaderConnecting(callbackContext);
    } else if (action.equals("onSwipeDetected")) {
      handler.onSwipeDetected(callbackContext);
    } else if (action.equals("onBatteryLow")) {
      handler.onBatteryLow(callbackContext);
    } else if (action.equals("onReaderDisconnected")) {
      handler.onReaderDisconnected(callbackContext);
    } else if (action.equals("tokenizeLastSwipe")) {
      handler.tokenizeCard(callbackContext);
    } else if (action.equals("onReaderConnected")) {
      handler.onReaderConnected(callbackContext);
    } else if (action.equals("watchForSwipe")) {
      this.watchForSwipe(callbackContext);
    } else if (action.equals("onCardRead")) {
      handler.onCardRead(callbackContext);
    } else if (action.equals("startOnReaderAttached")) {
      handler.onReaderAttached(callbackContext);
    } else if (action.equals("onReaderFail")){
      handler.onReaderFail(callbackContext);
    } else if (action.equals("stopSwipe")) {
      reader.StopSwipe();
    } else {
      success = false;
    }
    
    return success;
  }

  private void authorizeCardFlightAccount(String apiToken, String stripeMerchantToken, CallbackContext callbackContext) {
    if (apiToken == null || stripeMerchantToken == null) {
      logError("Need to send both an api token and a stripe merchant token to authorize cardflight");
      callbackContext.error("Need to send both an api token and a stripe merchant token to authorize cardflight");
    }

    CardFlight.getInstance().setApiTokenAndAccountToken(apiToken, stripeMerchantToken);
    log("CardFlight authorized on this device");
    callbackContext.success("CardFlight authorized on this device");
  }

  private void initializeReader(final CallbackContext callbackContext) {
    log("CardFlight reader initializing");
    // cdv.getThreadPool().execute(new Runnable() {
    //   public void run() {
    //     reader = new Reader(cdv.getActivity().getApplicationContext(), handler);
    //     callbackContext.success("CardFlight reader initialized");
    //     // reader = new Reader(this.cordova.getActivity().getApplicationContext(), handler, new AutoConfigHandler(callbackContext));
    //     // log("Reader AutoConfigHandler has cordova callback");
    //   }
    // });

    reader = new Reader(cdv.getActivity().getApplicationContext(), handler);
    this.handler.setReader(reader);
    callbackContext.success("CardFlight reader initialized");
  }

  private void watchForSwipe(CallbackContext callbackContext) {
    handler.resetCard();
    reader.beginSwipe();
    log("CardFlight reader awaiting swipe");
    callbackContext.success("CardFlight reader awaiting swipe");
  }

  private void log(String s) {
    Log.i("CDVCardFlight", s);
  }

  private void logError(String s) {
    Log.e("CDVCardFlight", s);
  }
}